Add Compile and Release Workflow

28Smiles opened the pull request:
Adds a workflow building the app for the web and compiling with tauri on macos, windows and linux. Those artifacts are automatically attached to a prerelease with tag `latest`.
Dont forget to squash.

28Smiles committed:
Create build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Merge branch 'fh-igd-iet:trunk' into trunk

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+name: Build
+
+on:
+  push:
```
It may be better to run on pull requests as well to validate them

28Smiles replied:
Afaik, this already includes pull requests

fh-igd-mueller-roemer replied:
I'm no expert on GitHub actions, but the documentation makes it sound like `pull_request` is a different trigger: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request

28Smiles replied:
I tested it with this PR (https://github.com/28Smiles/qcvis/pull/1), but it is possible, that the `pull_request` will use the trunks workflow and the `push` trigger the pr's workflow.

fh-igd-mueller-roemer replied:
Ok, I guess we can leave this "let's wait and see" and adjust if necessary

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+          name: tauri-web
+          path: dist/*
+  build-linux:
+    runs-on: ubuntu-latest
```
Ideally, we want the oldest supported Ubuntu LTS for widest compatibility of the generated executable

28Smiles replied:
I will change this

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+      - uses: actions/setup-node@v1
+        name: Setup Node
+        with:
+          node-version: "*"
```
I would prefer building with LTS-releases, i.e., `lts/*`. Additionally, would it make sense to activate `cache: yarn` here and in the following `setup-node` calls?

28Smiles replied:
Should we build with multiple versions of node? This should not be necessary since we don't use node as a runtime, merely as a build tool (which should create the same artifact regardless of the node version).

fh-igd-mueller-roemer replied:
I think a single version (newest LTS) is sufficient in this case

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+      - name: Install Yarn
+        run: corepack enable
+      - name: Install npm dependencies
+        run: yarn install
```
Is `--immutable` implied in GitHub actions?

28Smiles replied:
I will add this

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+      - name: Install npm dependencies
+        run: yarn install
+      - name: Build
+        run: yarn run build
```
Does it make sense (and is it possible) to build the developer version as well, to validate that it also works?

28Smiles replied:
there is no developer version, the developer version is only served by a dev server in development. We could build a dev version, but it is not currently intended by the npm build config

fh-igd-mueller-roemer replied:
At the very least things like `import.meta.env.PROD` are different, and I believe optimization is also omitted

fh-igd-mueller-roemer replied:
But I think we can leave this be for now

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+      - name: Install npm dependencies
+        run: yarn install
+      - name: Install tauri dependencies
+        run: sudo apt update && sudo apt install -y build-essential libpython3-dev libdbus-1-dev libgtk-3-dev libsoup-gnome2.4-dev libwebkit2gtk-4.0-dev
```
Is `build-essential` really necessary? Also stylistically, I would prefer a list of commands, i.e.,

```yaml
run: |
  sudo apt update
  sudo apt install -y # ...
```

I don't know if it's any good, but `awalsh128/cache-apt-pkgs-action` may be a better approach (the list of packages installed here is long, especially considering the transitive dependencies)

28Smiles replied:
I will look into it

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+        name: Clone
+      - uses: actions-rs/toolchain@v1
+        name: Setup Rust
+        with:
```
Wouldn't the minimal profile be sufficient? Is the toolchain cached?

28Smiles replied:
I don't know if it is cached (should be done by microsoft). There is a version of rust preinstalled in the runner image, but for consistency and newest updates it is practice to install it manually.

fh-igd-mueller-roemer replied:
Definitely, I meant the install files. `actions-rs` isn't directly by GitHub/Microsoft AFAIK. `setup-node` takes care of this internally, so probably `toolchain` does so too

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+        with:
+          name: tauri-debian
+          path: src-tauri/target/release/bundle/deb/*.deb
+  build-windows:
```
Can these (`build-linux`, `build-windows`, `build-macos`) all be bundled via an OS matrix? The uploads are different, but I think `upload-artifact` also supports multiple paths (and name can be parametrized on the OS)

28Smiles replied:
Yes, i just preferred it this way, cause of the different artifact count and packages needed, but with some conditional tasks, this could be modeled as an os-matrix

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+      - name: Build Tauri
+        run: yarn run build:tauri
+      - name: Compress App
+        run: tar -cvzf qcvis.app.tar.gz ./src-tauri/target/release/bundle/macos/qcvis.app
```
Doesn't `upload-artifact` itself perform compression?

28Smiles replied:
yes, but since `qcvis.app` is a folder we want to include, we wrap everything in a tar to attach it to the release

fh-igd-mueller-roemer replied:
I was still mostly thinking about the OS matrix in this case. The `release`-stage (which I want to omit right now) could do the final bundling

28Smiles replied:
I had problems, uploading the artifact, retaining the folder structure.

fh-igd-mueller-roemer replied:
Ok then leave it as is (with a conditional if you switch to an OS matrix)

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+        with:
+          name: tauri-macos-app
+          path: qcvis.app.tar.gz
+  release:
```
There are a few things we need to clear up before we can do automated releases, so I'd prefer to defer this for now

28Smiles replied:
This is only a development (pre-)release, releasing the latest trunk, overriding the release each time. This is not planned not supposed to be a tagged release.

fh-igd-mueller-roemer replied:
I know, but there are still some issues we need to clear up first before we do pre-bundled releases (no matter if pre-release or tagged), so I'd like to see this as a separate PR

28Smiles replied:
Okay, I will deactivate the release task then.

What kind of blocking issues are there?

fh-igd-mueller-roemer replied:
Mostly regarding the documentation of dependencies in the released versions. I have someone working on it

fh-igd-mueller-roemer replied:
Not 100% happy with our solution (#4) yet, but it'll do for now

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,172 @@
+      - name: Install tauri dependencies
+        run: sudo apt update && sudo apt install -y build-essential libpython3-dev libdbus-1-dev libgtk-3-dev libsoup-gnome2.4-dev libwebkit2gtk-4.0-dev
+      - name: Build Tauri
+        run: yarn run build:tauri
```
This internally calls Rust. Due to Tauri's many dependencies, we may want to use `Swatinem/rust-cache@v1`

28Smiles replied:
Yes we can add a cache, but the build process is not bottlenecked by the dependency downloads, but by building them. We could (also) cache the build libraries, but I think a clean build would be preferable, since rust and llvm versions may change, resulting in worse binaries.

fh-igd-mueller-roemer replied:
`rust-cache` caches dependency target files as well. A different version would result in a re-build anyway

fh-igd-mueller-roemer replied:
And the rust/llvm version info is added to the cache key automatically, so no issues there

fh-igd-mueller-roemer replied:
One more addition here: just like `yarn install` should be run with a fixed .lock file, the same should be true for `cargo build`. This can be done via `yarn run build:tauri -- --locked`

fh-igd-mueller-roemer reviewed (commented):
Since this is only a configuration file, not part of the actual source code, and GitHub-specific, we can probably merge this without a CLA. However I have a few comments / questions

28Smiles commented:
To further reduce binary size we could compress the executeables with upx, but I dont think it will be necessary.

fh-igd-mueller-roemer commented:
We definitely don't want to use UPX

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

28Smiles committed:
Update build.yml

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -8,27 +8,29 @@ env:
-          node-version: "*"
-      - name: Install Yarn
-        run: corepack enable
+          node-version: 16
```
Why 16 not `lts/*`? I'd be fine with using the latest LTS (although this is ok too)

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -8,27 +8,29 @@ env:
 jobs:
   build-web:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
```
For the web version the Ubuntu version shouldn't matter, it's only about binaries

fh-igd-mueller-roemer commented on .github/workflows/build.yml:
```diff
@@ -37,81 +39,51 @@ jobs:
           path: src-tauri/target/release/bundle/appimage/*.AppImage
-      - uses: actions/upload-artifact@v3
+      - if: matrix.os == 'ubuntu-20.04'
+        uses: actions/upload-artifact@v3
```
Can't the uploads be bundled into a single one with multiple globs and a parametric name?

28Smiles replied:
we'd still need one artifact for each matrix run (3 to be clear). So its possible, but we would need to introduce more parameters for the matrix run or use the os type as artifact name (ugly imo). Also there is still the Problem with the MacOS app directory artifact.

fh-igd-mueller-roemer replied:
I don't have an issue with using the OS name in the artifact name. For macOS, the directory compression can be done conditionally. I'm not against conditionals in general, I just want to minimize them here if possible

fh-igd-mueller-roemer reviewed (commented):
Some new comments / questions

28Smiles committed:
Update build.yml

fh-igd-mueller-roemer commented:
@28Smiles looking at the build logs, I think you still need to set the Rust workspace correctly for `Swatinem/rust-cache`. From the post step:

```
Warning:  Command failed: cargo metadata --all-features --format-version 1
Warning:  error: could not find `Cargo.toml` in `/home/runner/work/qcvis/qcvis` or any parent directory

... Cleaning /home/runner/work/qcvis/qcvis/target ...
Warning:  Error: ENOENT: no such file or directory, opendir '/home/runner/work/qcvis/qcvis/target'
... Cleaning cargo registry ...
... Cleaning cargo/bin ...
... Cleaning cargo git cache ...
Warning:  Error: ENOENT: no such file or directory, opendir '/home/runner/.cargo/git/db'
... Saving cache ..
```

fh-igd-mueller-roemer committed:
Match yarn dependency format for tauri (#3)

Produce a raw text file that matches the Yarn dependency plugin output format by adjusting `about.hbs`

fh-igd-mueller-roemer reviewed (commented):
Addition regarding the Tauri builds

fh-igd-mueller-roemer committed:
Add license text and attribution notices to output (#4)

Co-authored-by: fh-igd-lotter <lars.lotter@igd.fraunhofer.de>

fh-igd-mueller-roemer committed:
Add build.yml with lints

fh-igd-mueller-roemer committed:
workspaces -> working-directory

fh-igd-mueller-roemer committed:
Does this work?

fh-igd-mueller-roemer committed:
Add rustfmt since we use it

fh-igd-mueller-roemer committed:
Updates and fixes

fh-igd-mueller-roemer committed:
Create dummy dist directory during linking

fh-igd-mueller-roemer committed:
Add back workspace (didn't work without either)

fh-igd-mueller-roemer committed:
Specify lts/hydrogen

fh-igd-mueller-roemer committed:
Merge remote-tracking branch '28Smiles/trunk' into lints

Co-authored-by: 28Smiles <Leon.c@gmx.de>

fh-igd-mueller-roemer committed:
Cleanup

fh-igd-mueller-roemer committed:
Use locked builds

fh-igd-mueller-roemer commented:
@28Smiles I am currently merging and updating this over in https://github.com/fh-igd-mueller-roemer/qcvis/tree/lints

fh-igd-mueller-roemer committed:
Simplify artifact uploads and downloads

fh-igd-mueller-roemer committed:
Move -- --locked to build.yml

fh-igd-mueller-roemer committed:
Consitstent naming

fh-igd-mueller-roemer committed:
Attempt to fix release
